File: lab3/msLab3/src/classes/Process.ts
import { Element } from "./Element.js";
import { Job } from "./Job.js";

export class Process extends Element {
  protected queue: Job[] = [];
  protected channels: Process.Channel[] = [];
  protected failures: number = 0;
  protected maxQueueSize: number = Number.MAX_SAFE_INTEGER;
  protected meanQueue: number = 0.0;
  protected workTime: number = 0.0;
  protected totalLeaveTime: number = 0.0;
  protected previousLeaveTime: number = 0.0;

  constructor(
    name: string,
    delayMean: number,
    channelsNum: number,
    delayDev?: number
  ) {
    super(name, delayMean, delayDev);
    for (let i = 0; i < channelsNum; i++) {
      this.channels.push(new Process.Channel());
    }
  }

  public initializeChannelsWithJobs(jobsNum: number): void {
    const limitedJobsNum = Math.min(jobsNum, this.channels.length);
    for (let i = 0; i < limitedJobsNum; i++) {
      this.channels[i].setCurrentJob(new Job(0.0));
      this.channels[i].setTNext(this.getTCurr() + this.getDelay());
    }
  }

  public initializeQueueWithJobs(jobsNum: number): void {
    const limitedJobsNum = Math.min(jobsNum, this.maxQueueSize);
    for (let i = 0; i < limitedJobsNum; i++) {
      this.queue.push(new Job(0.0));
    }
  }

  public override inAct(job: Job): void {
    const freeChannel = this.getFreeChannel();
    if (freeChannel) {
      freeChannel.setCurrentJob(job);
      freeChannel.setTNext(this.getTCurr() + this.getDelay());
    } else {
      if (this.queue.length < this.getMaxQueueSize()) {
        this.queue.push(job);
      } else {
        this.failures++;
      }
    }
  }

  public override outAct(): void {
    this.processCurrentJobs();
    this.startNextJobs();
  }

  protected processCurrentJobs(): void {
    const channelsWithMinTNext = this.getChannelsWithMinTNext();
    for (const channel of channelsWithMinTNext) {
      const job = channel.getCurrentJob();

      if (!job) continue;

      const nextRoute = this.getNextRoute(job);
      if (nextRoute.isBlocked(job)) continue;

      if (nextRoute.getElement()) {
        job.setTimeOut(this.getTCurr());
        nextRoute.getElement()?.inAct(job);
      }

      channel.setCurrentJob(null);
      channel.setTNext(Number.MAX_VALUE);
      this.changeQuantity(1);
      this.totalLeaveTime += this.getTCurr() - this.previousLeaveTime;
      this.previousLeaveTime = this.getTCurr();
    }
  }

  protected startNextJobs(): void {
    let freeChannel = this.getFreeChannel();
    while (this.queue.length > 0 && freeChannel) {
      const job = this.queue.shift();
      if (job) {
        freeChannel.setCurrentJob(job);
        freeChannel.setTNext(this.getTCurr() + this.getDelay());
      }
      freeChannel = this.getFreeChannel();
    }
  }

  protected getChannelsWithMinTNext(): Process.Channel[] {
    const channelsWithMinTNext: Process.Channel[] = [];
    let minTNext = Number.MAX_VALUE;

    for (const channel of this.channels) {
      if (channel.getTNext() < minTNext) {
        minTNext = channel.getTNext();
      }
    }

    for (const channel of this.channels) {
      if (channel.getTNext() === minTNext) {
        channelsWithMinTNext.push(channel);
      }
    }

    return channelsWithMinTNext;
  }

  protected getFreeChannel(): Process.Channel | null {
    for (const channel of this.channels) {
      if (channel.getState() === 0) {
        return channel;
      }
    }
    return null;
  }

  public getMaxQueueSize(): number {
    return this.maxQueueSize;
  }

  public setMaxQueueSize(maxQueueSize: number): void {
    this.maxQueueSize = maxQueueSize;
  }

  public getFailures(): number {
    return this.failures;
  }

  public getMeanQueue(): number {
    return this.meanQueue;
  }

  public getWorkTime(): number {
    return this.workTime;
  }

  public override doStatistics(delta: number): void {
    super.doStatistics(delta);
    this.meanQueue += this.queue.length * delta;
    this.workTime += this.getState() * delta;
  }

  public override getState(): number {
    return this.channels.reduce(
      (state, channel) => state | channel.getState(),
      0
    );
  }

  public override getTNext(): number {
    return Math.min(...this.channels.map((channel) => channel.getTNext()));
  }

  public override setTNext(tNext: number): void {
    const previousTNext = this.getTNext();
    for (const channel of this.channels) {
      if (channel.getTNext() === previousTNext) {
        channel.setTNext(tNext);
      }
    }
  }

  public override printInfo(): void {
    console.log(
      `${this.getName()} state = ${this.getState()} quantity = ${this.getQuantity()} tnext = ${this.getTNext()} failures = ${
        this.failures
      } queue size = ${this.queue.length}`
    );
  }

  public getQueueSize(): number {
    return this.queue.length;
  }

  public getMeanLeaveInterval(): number {
    return this.totalLeaveTime / this.getQuantity();
  }

  public getUnprocessedJobs(): Job[] {
    const jobs: Job[] = [];
    for (const channel of this.channels) {
      if (channel.getCurrentJob()) {
        jobs.push(channel.getCurrentJob());
      }
    }
    if (this.queue.length > 0) {
      jobs.push(...this.queue);
    }
    for (const job of jobs) {
      job.setTimeOut(this.getTCurr());
    }
    return jobs;
  }
}

export namespace Process {
  export class Channel {
    private currentJob: Job | null = null;
    private tNext: number = Number.MAX_VALUE;

    public getState(): number {
      return this.currentJob === null ? 0 : 1;
    }

    public getCurrentJob(): Job | null {
      return this.currentJob;
    }

    public setCurrentJob(job: Job | null): void {
      this.currentJob = job;
    }

    public getTNext(): number {
      return this.tNext;
    }

    public setTNext(tNext: number): void {
      this.tNext = tNext;
    }
  }
}

File: lab1/pythonProject/core/core/Process.py
from collections import deque
from core.core.Element import Element
from core.core.Job import Job

class Process(Element):
    def __init__(self, name, delay_mean, channels_num, delay_dev=None):
        super().__init__(name, delay_mean, delay_dev)
        self.queue = deque()
        self.channels = [self.Channel() for _ in range(channels_num)]
        self.failures = 0
        self.max_queue_size = float('inf')
        self.mean_queue = 0.0
        self.work_time = 0.0
        self.total_leave_time = 0.0
        self.previous_leave_time = 0.0

    def initialize_channels_with_jobs(self, jobs_num):
        jobs_num = min(jobs_num, len(self.channels))
        for i in range(jobs_num):
            self.channels[i].set_current_job(Job(0.0))
            self.channels[i].set_t_next(self.get_tcurr() + self.get_delay())

    def initialize_queue_with_jobs(self, jobs_num):
        jobs_num = min(jobs_num, self.max_queue_size)
        for i in range(jobs_num):
            self.queue.append(Job(0.0))

    def in_act(self, job):
        free_channel = self.get_free_channel()
        if free_channel is not None:
            free_channel.set_current_job(job)
            free_channel.set_t_next(self.get_tcurr() + self.get_delay())
        else:
            if len(self.queue) < self.max_queue_size:
                self.queue.append(job)
            else:
                self.failures += 1

    def out_act(self):
        self.process_current_jobs()
        self.start_next_jobs()

    def process_current_jobs(self):
        channels_with_min_tnext = self.get_channels_with_min_tnext()
        for channel in channels_with_min_tnext:
            job = channel.get_current_job()
            next_route = self.get_next_route(job)
            if next_route.is_blocked(job):
                continue
            if next_route.get_element() is not None:
                job.set_time_out(self.get_tcurr())
                next_route.get_element().in_act(job)
            channel.set_current_job(None)
            channel.set_t_next(float('inf'))
            self.change_quantity(1)
            self.total_leave_time += self.get_tcurr() - self.previous_leave_time
            self.previous_leave_time = self.get_tcurr()

    def start_next_jobs(self):
        free_channel = self.get_free_channel()
        while self.queue and free_channel is not None:
            job = self.queue.popleft()
            free_channel.set_current_job(job)
            free_channel.set_t_next(self.get_tcurr() + self.get_delay())
            free_channel = self.get_free_channel()

    def get_channels_with_min_tnext(self):
        min_tnext = float('inf')
        channels_with_min_tnext = []
        for channel in self.channels:
            if channel.get_t_next() < min_tnext:
                min_tnext = channel.get_t_next()
        for channel in self.channels:
            if channel.get_t_next() == min_tnext:
                channels_with_min_tnext.append(channel)
        return channels_with_min_tnext

    def get_free_channel(self):
        for channel in self.channels:
            if channel.get_state() == 0:
                return channel
        return None

    def do_statistics(self, delta):
        super().do_statistics(delta)
        self.mean_queue += len(self.queue) * delta
        self.work_time += self.get_state() * delta

    def get_state(self):
        state = 0
        for channel in self.channels:
            state |= channel.get_state()
        return state

    def get_t_next(self):
        t_next = float('inf')
        for channel in self.channels:
            if channel.get_t_next() < t_next:
                t_next = channel.get_t_next()
        return t_next

    def set_t_next(self, t_next):
        previous_tnext = self.get_t_next()
        for channel in self.channels:
            if channel.get_t_next() == previous_tnext:
                channel.set_t_next(t_next)

    def print_info(self):
        print(f"{self.get_name()} state = {self.get_state()} quantity = {self.get_quantity()} tnext = {self.get_t_next()} failures = {self.failures} queue size = {len(self.queue)}")

    def get_queue_size(self):
        return len(self.queue)

    def get_mean_leave_interval(self):
        return self.total_leave_time / self.get_quantity()

    def get_unprocessed_jobs(self):
        jobs = [channel.get_current_job() for channel in self.channels if channel.get_current_job() is not None]
        jobs.extend(self.queue)
        for job in jobs:
            job.set_time_out(self.get_tcurr())
        return jobs

    class Channel:
        def __init__(self):
            self.current_job = None
            self.t_next = float('inf')

        def get_state(self):
            return 1 if self.current_job is not None else 0

        def get_current_job(self):
            return self.current_job

        def set_current_job(self, job):
            self.current_job = job

        def get_t_next(self):
            return self.t_next

        def set_t_next(self, t_next):
            self.t_next = t_next
